pkg/controller/volume/attachdetach/util/util.go

```diff
@@ -66,6 +71,25 @@ func CreateVolumeSpec(podVolume v1.Volume, podNamespace string, pvcLister coreli
 				err)
 		}
 
+		translate, err := isCSITranslationNeeded(volumeSpec, nodeName, vpm, csiMigratedPluginManager)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"error checking CSI migration status of volume plugin for PVC %q/%q: %v",
+				podNamespace,
+				pvcSource.ClaimName,
+				err)
+		}
+		if translate {
+			volumeSpec, err = volume.TranslateInTreeSpecToCSI(volumeSpec, csiTranslator)
+			if err != nil {
+				return nil, fmt.Errorf(
+					"error translating volume spec for PVC %q/%q for CSI migration: %v",
+					podNamespace,
+					pvcSource.ClaimName,
+					err)
+			}
+		}
+
```

Comment:
This pattern repeats several times, can it be moved to a separate function / csiMigratedPluginManager method?

Reply:
Note that the logic in `isCSITranslationNeeded` is a bit unique in that only the AttachDetach controller takes the extra `isCSIMigrationSupportedOnNode` logic into account. If that logic is necessary in other controllers, we can consider moving it to the CSIMigratedPluginManager. For now I can create a utility function in AttachDetach controller to perform the check-translation-necessary-and-translate logic above.